🎨(app) make nginx an app-specific service

jmaupetit opened the pull request:
## Purpose

Until now, we considered that nginx may be a single service used by many apps at the same time. Even if this pattern is quite reasonable, it offers less configurability if we want to deploy not all but a selection of applications Arnold can play with. Hence we decided to split the monolithic nginx service into multiple nginx services (one per application). 

## Proposal

- [x] split the common/nginx svc
- [x] split the common/nginx dc
- [x] split nginx configMaps
- [x] redefine app-specific routes to point to the appropriate nginx svc
- [x] ~improve route aliases definition to be app specific~
- [x] add a `redirect` app whose responsibility is to handle route aliases
- [x] move ES common service to richie

jmaupetit committed:
🎨(app) make nginx an app-specific service

Until now, we considered that nginx may be a single service used by many
apps at the same time. Even if this pattern is quite reasonable, it
offers less configurability if we want to deploy not all but a selection
of applications Arnold can play with. Hence we decided to split the
monolithic nginx service into multiple nginx services (one per
application). It means that we had to:

* split the common/nginx svc
* split the common/nginx dc
* split nginx configMaps
* redefine app-specific routes to point to the appropriate nginx svc
* improve route aliases definition to be app specific

sampaccoud commented on group_vars/all/main.yml:
```diff
@@ -82,11 +82,12 @@ richie_database:
 
 # Openshift template files
 openshift_deployments:
+  - openshift/common/dc/elasticsearch.yml
```
In fact it's not common because richie and edxapp require different versions of ES. So this one is dedicated to Richie.

jmaupetit replied:
True. For now, I will make this dc/svc richie-specific. We will add ES for edxapp later.

cchaudier replied:
I think we have to abandon the `common` directory.
I think we should consider the `templates/openshift` directory as a collection of usable micro-services for an application.
I propose that we adopt the following naming convention:
`templates/openshift/<micro-service-name>/{svc,dc,pcv,...}`

If we do like this in the future we can find all templates use by a microservice and not hardcode the path in an ansible vars.

jmaupetit replied:
Agreed! I'll switch to this pattern next week.

sampaccoud commented on templates/openshift/common/dc/redirect.yml.j2:
```diff
@@ -0,0 +1,33 @@
+apiVersion: v1
```
I was expecting a "redirect" directory at the same level as "edxapp" and "richie" but the svc/dc would still be nginx:

 templates/openshift/redirect/dc/nginx.yml.j2
 templates/openshift/redirect/svc/nginx.yml.j2

WDYT?

jmaupetit replied:
Totally right. Let's do this.

cchaudier replied:
It's a very good idea and close to the naming convention I propose.

jmaupetit committed:
🎨(app) move ES from common to richie

The currently used image of ElasticSearch is only compatible with Richie
(and not edxapp). So let's make it a richie dc/svc.

cchaudier commented on templates/openshift/edxapp/dc/nginx.yml.j2:
```diff
@@ -0,0 +1,47 @@
+kind: DeploymentConfig
+metadata:
+  labels:
+    app: nginx
```
All version of an app are grouped on th same app on OpenShift ?
Why do not include the deployment_stamp also ?
The app must be `edxapp` no ?

jmaupetit replied:
Yes! and that will be part of my next refactoring. IMO this is not part of the scope of this PR.

cchaudier replied:
Ok

cchaudier commented on templates/openshift/common/svc/redirect.yml.j2:
```diff
@@ -0,0 +1,19 @@
+  namespace: "{{ project_name }}"
+spec:
+  ports:
+{% for port in nginx_ports %}
```
If we split nigix into app micro-service whe don't need the nginx_ports and this loop.

jmaupetit replied:
Yes. I'll do this refactoring next week.

cchaudier replied:
ok

cchaudier reviewed (changes requested):
Some question and proposition.
Great job !
